### PR TITLE
[Nathanael Seen Zhong Qi] Duke Increments

### PR DESCRIPTION
**Pull request by nathanaelseen**

### CODE REVIEW

**Review by oLiXino — COMMENTED** [review]
The implementation of enum improves the readability of the codes and avoid errors from passing in invalid constants. The code is clean and readable.

**Review by oLiXino — COMMENTED** [review]
The separation of code of the Duke class into Ui class complies with the OOP paradigm and makes the code look cleaner and more readable.

File: src/main/java/duke/Command.java [hunk 1]
```diff
@@ -16,4 +16,4 @@ public CommandType getCommandType() {
     public String[] getParams() {
         return params;
```
Comment (loycatherine) [inline src/main/java/duke/Command.java]: Maybe you could return this.params instead?

File: src/main/java/duke/Duke.java [hunk 2]
```diff
@@ -43,49 +40,43 @@ public void run() throws IOException {
...
                     case DEADLINE_CMD:
                         t = new Deadline(command.getParams()[0], LocalDate.parse(command.getParams()[1]));
                         addAndPrintTask(t, taskList);
+
                         storage.saveTasksToFile(taskList);
                         break;
                     case EVENT_CMD:
                         t = new Event(command.getParams()[0], LocalDate.parse(command.getParams()[1]));
                         addAndPrintTask(t, taskList);
+
```
Comment (loycatherine) [inline src/main/java/duke/Duke.java]: In this module, when you use the switch, try to do it like this:
switch (command.getCommandType()) {
case EVENT_CMD: 
...
}

File: src/main/java/duke/Command.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package duke;
+
+/**
+ * Represents a Command entered by the user.
+ */
+public class Command {
+    private CommandType commandType;
```
Comment (RubaP) [inline src/main/java/duke/Command.java:7]: Think what if you implement subclasses each one for single Specific Command.  What kind of benefits you would get? You can discuss this with your team as well for your team project.

File: src/main/java/duke/CommandType.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package duke;
+
+/**
+ * Represents the type of command entered by the user.
+ */
+public enum CommandType {
+    LIST_CMD("list"),
```
Comment (RubaP) [inline src/main/java/duke/CommandType.java:7]: Good that you have defined all the commands as enum

File: src/main/java/duke/Duke.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    /**
+     * Constructs a new Duke object.
+     *
+     */
+    public Duke() {
+        try {
+            ui = new Ui();
+            parser = new Parser();
+
+            storage = new Storage(Storage.DUKE_TXT_FILE_PATH);
```
Comment (RubaP) [inline src/main/java/duke/Duke.java:24]: Good that you have defined file path as constant

File: src/main/java/duke/DukeException.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+ * Represents any error that occurs in the context of Duke (excluding IO application).
+ */
+public class DukeException extends Exception {
+    /**
+     * Constructs a new Duke Exception object.
+     *
+     * @param s The message of the error.
+     */
+    public DukeException(String s) {
+        // Call constructor of parent Exception
```
Comment (RubaP) [inline src/main/java/duke/DukeException.java]: You can avoid trivial comments

File: src/main/java/duke/Main.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    private Duke duke;
+
+    /**
+     * Creates a Main instance object.
+     */
+    public Main() {
+        //try {
+            duke = new Duke();
+        /*} catch (IOException e) {
```
Comment (RubaP) [inline src/main/java/duke/Main.java]: Remove commented code blocks

File: src/main/java/duke/Parser.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            id = Integer.parseInt(strId);
+        } catch (NumberFormatException e) {
+            return false;
+        }
+
+        return (!((id - 1) > tasks.getNumTasks() - 1 || (id - 1) < 0));
+    }
+
+    /**
+     * Does the main work of parsing raw commands from user input.
```
Comment (RubaP) [inline src/main/java/duke/Parser.java]: Method comment should start with simple present e.g. Parses xxx

File: src/main/java/duke/Parser.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+     *
+     * @param cmd The raw command entered by user.
+     * @param tasks The current task-list.
+     * @return Command object representing the parsed user command.
+     * @throws DukeException If the command is invalid or if there were errors parsing the command.
+     */
+    public Command parse(String cmd, TaskList tasks) throws DukeException {
```
Comment (RubaP) [inline src/main/java/duke/Parser.java:53]: Your parse method is very lengthy which will make it difficult to understand. Think of possible ways to improve the readability of the code. E.g. extracting out code segments as methods

File: src/main/java/duke/Storage.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+     * Constructs a new Storage object. Opens the file if it exists, else creates a new one.
+     *
+     * @param filePath The relative file path of the data (text) file on the computer.
+     * @throws IOException If there are any IO-related exceptions while trying to open or create a file.
+     */
+    public Storage(String filePath) throws IOException {
+        file = new File(filePath);
+
+        if (!file.exists()) {
+            // Create file if it does not already exists
```
Comment (RubaP) [inline src/main/java/duke/Storage.java]: You can avoid trivial comments

File: src/main/java/duke/TaskList.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
+package duke;
+
+import java.util.List;
+import java.util.ArrayList;
+
+/**
+ * Represents the task-list containing all Tasks the user would record.
+ */
+public class TaskList {
```
Comment (RubaP) [inline src/main/java/duke/TaskList.java:9]: Very good design of TaskList

File: src/main/java/duke/Ui.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+import java.io.IOException;
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+
+/**
+ * Performs basic 'interaction' with the user.
+ */
+public class Ui {
+    public static final String LF = "\r\n";
```
Comment (RubaP) [inline src/main/java/duke/Ui.java:9]: Good that you have defined the messages as constants

**Review by RubaP — COMMENTED** [review]
Overall you have done a great job! Please go through the detailed comments given. For the design level comments, you can discuss it with your team what others have done what are pros and cons of different approaches. You can also focus on implementing more Junit test cases

**Pull request closed without merging**